MaxText/layers/linears.py

```diff
@@ -635,49 +720,85 @@ def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
     weights = self.reshape_and_update_weights(top_k_weights, top_k_indices)
     matmul_precision = lax.Precision(self.config.matmul_precision)
 
+    if self.config.model_call_mode != "inference":
+      loss = self.load_balance_loss(top_k_indices, weights)
+    else:
+      loss = None
+    batch_size = inputs.shape[0]
+    seq_len = inputs.shape[1]
+
+    cp, sub_seq = self.get_context_partition_and_sub_seq(seq_len)
+
     if self.config.capacity_factor > 0:
       # token dropping if needed
-      dispatch_mask, combine_mask = self.generate_masks(top_k_indices, weights)
-      mask_axes = ("activation_batch", "activation_length", None, None)
-      dispatch_mask = nn.with_logical_constraint(dispatch_mask, mask_axes)
-      combine_mask = nn.with_logical_constraint(combine_mask, mask_axes)
       if self.config.model_call_mode != "inference":
-        loss = self.load_balance_loss(top_k_indices, weights)
+        dispatch_mask, combine_mask = self.generate_masks(top_k_indices, weights)
+        mask_axes = ("activation_batch", "activation_length", None, None)
+        input_axis = ("activation_batch", "activation_length", "activation_embed")
+        dispatch_axis = ("activation_exp", "activation_batch_no_exp", None, "activation_embed")
+        mlp_axis = ("activation_exp", "activation_batch_no_exp", None, "activation_mlp")
+        dispatch_eimsum = f"BSM,BSEC -> EBCM"
+        mlp_einsum = f"EBCM,EMH -> EBCH"
+        output_einsum = f"EBCM,BSEC -> BSM"
       else:
-        loss = None
-      inputs = nn.with_logical_constraint(inputs, ("activation_batch", "activation_length", "activation_embed"))
+        # todo: try replace softmax_probs with padded weights and verify with decode acc tests 
+        softmax_probs = jax.nn.softmax(gate_logits.astype(jnp.float32), axis=-1).astype(self.dtype)
```

Comment:
currently it is branched for softmax weight for training/inference. Ideally we don't like branching too much and later on the tests couldn't catch those branch changes. I added todo comments. later on next week to test out the accuracy using padded weight, we can remove that part of  branched code.